File: backend/controllers/admin_user_controller.py
from fastapi import APIRouter, HTTPException, Depends, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from models.schemas import AdminUser, AdminUserCreate, AdminUserUpdate
from services.auth_service import auth_service
from motor.motor_asyncio import AsyncIOMotorDatabase
from typing import List
from datetime import datetime, timezone
import uuid
import logging

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/admin-users", tags=["Admin Users"])
security = HTTPBearer()


async def verify_admin(credentials: HTTPAuthorizationCredentials, db: AsyncIOMotorDatabase) -> dict:
    """Verify admin user from token"""
    payload = auth_service.decode_token(credentials.credentials)
    if not payload or payload.get("type") != "access":
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid token")
    
    user = await db.admin_users.find_one({"id": payload["sub"]}, {"_id": 0})
    if not user or not user.get("is_active", True):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="User not found or disabled")
    
    return user


def create_admin_user_routes(db: AsyncIOMotorDatabase):
    """Create admin user management routes"""
    
    @router.get("/", response_model=List[AdminUser])
    async def list_admin_users(credentials: HTTPAuthorizationCredentials = Depends(security)):
        """List all admin users"""
        await verify_admin(credentials, db)
        users = await db.admin_users.find({}, {"_id": 0, "password_hash": 0}).to_list(100)
        return users
    
    @router.post("/", response_model=AdminUser, status_code=status.HTTP_201_CREATED)
    async def create_admin_user(
        user_data: AdminUserCreate,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Create a new admin user"""
        await verify_admin(credentials, db)
        
        # Check if username or email exists
        existing = await db.admin_users.find_one({
            "$or": [
                {"username": user_data.username},
                {"email": user_data.email}
            ]
        })
        if existing:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Username or email already exists"
            )
        
        # Create user
        user_id = str(uuid.uuid4())
        user_doc = {
            "id": user_id,
            "username": user_data.username,
            "email": user_data.email,
            "password_hash": auth_service.hash_password(user_data.password),
            "is_active": True,
            "created_at": datetime.now(timezone.utc).isoformat()
        }
        
        await db.admin_users.insert_one(user_doc)
        logger.info(f"Created admin user: {user_data.username}")
        
        # Return without password
        del user_doc["password_hash"]
        return user_doc
    
    @router.get("/{user_id}", response_model=AdminUser)
    async def get_admin_user(
        user_id: str,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Get a specific admin user"""
        await verify_admin(credentials, db)
        
        user = await db.admin_users.find_one({"id": user_id}, {"_id": 0, "password_hash": 0})
        if not user:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
        
        return user
    
    @router.put("/{user_id}", response_model=AdminUser)
    async def update_admin_user(
        user_id: str,
        user_data: AdminUserUpdate,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Update an admin user"""
        current_user = await verify_admin(credentials, db)
        
        # Check if user exists
        user = await db.admin_users.find_one({"id": user_id})
        if not user:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
        
        # Build update dict
        update_data = {}
        if user_data.username:
            # Check uniqueness
            existing = await db.admin_users.find_one({
                "username": user_data.username,
                "id": {"$ne": user_id}
            })
            if existing:
                raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Username already exists")
            update_data["username"] = user_data.username
        
        if user_data.email:
            existing = await db.admin_users.find_one({
                "email": user_data.email,
                "id": {"$ne": user_id}
            })
            if existing:
                raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Email already exists")
            update_data["email"] = user_data.email
        
        if user_data.password:
            update_data["password_hash"] = auth_service.hash_password(user_data.password)
        
        if user_data.is_active is not None:
            # Prevent disabling yourself
            if user_id == current_user["id"] and not user_data.is_active:
                raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Cannot disable your own account")
            update_data["is_active"] = user_data.is_active
        
        if update_data:
            await db.admin_users.update_one({"id": user_id}, {"$set": update_data})
        
        updated = await db.admin_users.find_one({"id": user_id}, {"_id": 0, "password_hash": 0})
        return updated
    
    @router.delete("/{user_id}", status_code=status.HTTP_204_NO_CONTENT)
    async def delete_admin_user(
        user_id: str,
        credentials: HTTPAuthorizationCredentials = Depends(security)
    ):
        """Delete an admin user"""
        current_user = await verify_admin(credentials, db)
        
        # Prevent self-deletion
        if user_id == current_user["id"]:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Cannot delete your own account")
        
        # Check if user exists
        user = await db.admin_users.find_one({"id": user_id})
        if not user:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
        
        await db.admin_users.delete_one({"id": user_id})
        logger.info(f"Deleted admin user: {user_id}")
    
    return router

File: backend/controllers/AdminUserController.js
/**
 * Admin User Controller
 * Single Responsibility: CRUD operations for admin users
 */

const bcrypt = require('bcryptjs');
const { v4: uuidv4 } = require('uuid');

class AdminUserController {
  constructor(collections) {
    this.collections = collections;
  }

  // GET /api/admin/users
  list = async (req, res) => {
    try {
      const users = await this.collections.adminUsers
        .find({}, { projection: { _id: 0, password_hash: 0 } })
        .toArray();

      res.json({
        success: true,
        users
      });
    } catch (error) {
      console.error('List users error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to list users'
      });
    }
  };

  // GET /api/admin/users/:id
  get = async (req, res) => {
    try {
      const { id } = req.params;

      const user = await this.collections.adminUsers.findOne(
        { id },
        { projection: { _id: 0, password_hash: 0 } }
      );

      if (!user) {
        return res.status(404).json({
          success: false,
          message: 'User not found'
        });
      }

      res.json({
        success: true,
        user
      });
    } catch (error) {
      console.error('Get user error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to get user'
      });
    }
  };

  // POST /api/admin/users
  create = async (req, res) => {
    try {
      const { username, email, password } = req.body;

      if (!username || !email || !password) {
        return res.status(400).json({
          success: false,
          message: 'Username, email, and password are required'
        });
      }

      // Check if username or email exists
      const existing = await this.collections.adminUsers.findOne({
        $or: [{ username }, { email }]
      });

      if (existing) {
        return res.status(400).json({
          success: false,
          message: 'Username or email already exists'
        });
      }

      const hashedPassword = await bcrypt.hash(password, 10);

      const user = {
        id: uuidv4(),
        username,
        email,
        password_hash: hashedPassword,
        is_active: true,
        created_at: new Date().toISOString()
      };

      await this.collections.adminUsers.insertOne(user);

      console.log(`✅ Created admin user: ${username}`);

      // Return without password
      const { password_hash, ...userResponse } = user;

      res.status(201).json({
        success: true,
        user: userResponse
      });
    } catch (error) {
      console.error('Create user error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to create user'
      });
    }
  };

  // PUT /api/admin/users/:id
  update = async (req, res) => {
    try {
      const { id } = req.params;
      const { username, email, password, is_active } = req.body;

      const user = await this.collections.adminUsers.findOne({ id });

      if (!user) {
        return res.status(404).json({
          success: false,
          message: 'User not found'
        });
      }

      const updateData = {};

      if (username && username !== user.username) {
        const existing = await this.collections.adminUsers.findOne({
          username,
          id: { $ne: id }
        });
        if (existing) {
          return res.status(400).json({
            success: false,
            message: 'Username already exists'
          });
        }
        updateData.username = username;
      }

      if (email && email !== user.email) {
        const existing = await this.collections.adminUsers.findOne({
          email,
          id: { $ne: id }
        });
        if (existing) {
          return res.status(400).json({
            success: false,
            message: 'Email already exists'
          });
        }
        updateData.email = email;
      }

      if (password) {
        updateData.password_hash = await bcrypt.hash(password, 10);
      }

      if (typeof is_active === 'boolean') {
        // Prevent self-deactivation
        if (id === req.user.id && !is_active) {
          return res.status(400).json({
            success: false,
            message: 'Cannot disable your own account'
          });
        }
        updateData.is_active = is_active;
      }

      if (Object.keys(updateData).length > 0) {
        await this.collections.adminUsers.updateOne(
          { id },
          { $set: updateData }
        );
      }

      const updated = await this.collections.adminUsers.findOne(
        { id },
        { projection: { _id: 0, password_hash: 0 } }
      );

      res.json({
        success: true,
        user: updated
      });
    } catch (error) {
      console.error('Update user error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to update user'
      });
    }
  };

  // DELETE /api/admin/users/:id
  delete = async (req, res) => {
    try {
      const { id } = req.params;

      // Prevent self-deletion
      if (id === req.user.id) {
        return res.status(400).json({
          success: false,
          message: 'Cannot delete your own account'
        });
      }

      const user = await this.collections.adminUsers.findOne({ id });

      if (!user) {
        return res.status(404).json({
          success: false,
          message: 'User not found'
        });
      }

      await this.collections.adminUsers.deleteOne({ id });

      console.log(`✅ Deleted admin user: ${id}`);

      res.json({
        success: true,
        message: 'User deleted'
      });
    } catch (error) {
      console.error('Delete user error:', error);
      res.status(500).json({
        success: false,
        message: 'Failed to delete user'
      });
    }
  };
}

module.exports = AdminUserController;
